web/modules/custom/joinup_notification/src/EntityMessageHelper.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_notification;
+
+use Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException;
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Entity\EntityStorageInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\message\MessageInterface;
+
+/**
+ * Service that assists in handling messages for entities.
+ *
+ * This is intended to be used for creating and retrieving messages that are
+ * related to entities.
+ *
+ * To send these messages, use the JoinupMessageDelivery service.
+ *
+ * There are different use cases for sending messages, this is only intended for
+ * basic creation and retrieval of messages. In case your entity requires custom
+ * business logic then feel free to inject this service and leverage it. See for
+ * example the InvitationMessageHelper which implements default arguments and
+ * resends a single message for each invitation instead of creating a new
+ * message every time.
+ *
+ * Because of the wild variations in business logic regarding messages this was
```

Comment:
I felt a small pat on my shoulder like someone was understanding me the moment I read the adjective 'wild' that described the word 'variations' of the notification messages' business logic.

Reply:
Yeah this was written after many hours of fruitless searching for a possible API that fits all use cases. Such a thing we will not see in our lifetimes.